functions/Set-DbaMaxMemoryEachInstanceBySize.ps1

```diff
@@ -0,0 +1,177 @@
+		foreach ($Service in $Services | Where-Object {$_.State -eq "Running"}) {
+			$Instance = $Server + "\" + $Service[1]
+			$Instance = $Instance.replace("\MSSQLSERVER", "").replace("MSSQL$","")
+
+			$Instances += ,@($Instance)
+		}
+
+		foreach ($Instance in $Instances) {
+			$Connection = Connect-DbaSqlServer -SqlServer $Instance
+
+			# Get memory on one server, no need to query it on each instance
+			if ($i -eq 0) {
+				$ServerMemoryMB = Get-DbaMaxMemory -SqlServer $Server
+
+				$ServerMemory = $ServerMemoryMB.TotalMB/1024
+
+				$i = 1
+			}
+
+			# Get total size of the user database data files
+			$Sql = "
+				SELECT SUM(CASE WHEN type = 0 THEN MF.size * 8 / 1024.0 /1024.0 ELSE 0 END)
+				FROM sys.master_files MF
+				JOIN sys.databases DB ON DB.database_id = MF.database_id
+				WHERE MF.database_id > 4 -- exclude system databases
+				  AND DB.source_database_id is null -- exclude snapshots;"
+    
+			$ServerSize = $Connection.ConnectionContext.ExecuteScalar($Sql)
+
+			$InstanceSizes += ,@($Instance, $ServerSize)
+		}
+
+		# Get the total size of all instances
+		foreach ($InstanceSize in $InstanceSizes) {
+			$TotalServerSize = $TotalServerSize + $InstanceSize[1]
+		}
+
+		# Jonathan Kehayias recommends to configure Max Server Memory based on the system configuration:
+		# Reserve 1 GB of RAM for the OS, 1 GB for each 4 GB of RAM installed from 4–16 GB, 
+		# and then 1 GB for every 8 GB RAM installed above 16 GB RAM. 
+		# Give what's left to SQL Server.
+		$Reserved = 1
+
+		if ($ServerMemory -ge 4) {
+			$Memory = $ServerMemory
+    
+			while ($Memory -gt 0) {
+				if ($Memory -ge 24) {
+					$Reserved += 1
+					$Memory = $Memory - 8
+				}
+				# Skip down to the 4GB-16GB calculation when between 17-23
+				elseif ($Memory -gt 16 -And $Memory -lt 24) {
+					$Memory = 16
+				}
+				else { #if ($Memory -le 16) 
+					$Reserved += 1
+					$Memory = $Memory - 4
+				}
+			}
+		}
+		else {
+			$Reserved = ($ServerMemory/2)
+		}
+
+		$TotalMaxMemorySize = $ServerMemory - $Reserved
+
+		# Sort the array by the instance sizes, smallest first
+		$InstanceSizes = $InstanceSizes | Sort-Object @{Expression={$_[1]}; Ascending=$True}
+
+		foreach ($InstanceSize in $InstanceSizes) {
+			# Give minimum of 2GB to each instance
+			if ($TotalMaxMemorySize * ($InstanceSize[1] / $TotalServerSize) -lt 2) {
+				$MaxMemorySize = 2
+
+				# Instance got more memory than it would have calculated to, so need to remove
+				# them from the totals
+				$TotalServerSize -= $InstanceSize[1]
+
+				$TotalMaxMemorySize -= 2
+        
+			}
+			# Get a percentage of the remaining memory based on instance size
+			else {
+				$MaxMemorySize = $TotalMaxMemorySize * ($InstanceSize[1] / $TotalServerSize)
+			}
+
+			$TotalMaxMemorySizeLeft = $TotalMaxMemorySizeLeft - $MaxMemorySize
+
+			# Max Server Memory is stored in megabytes
+			$MaxMemorySizeMb = $MaxMemorySize * 1024
+
+			Set-DbaMaxMemory -SqlServer $InstanceSize[0] -MaxMb $MaxMemorySizeMb
```

Comment:
Wouldn't it be best to check if there's a reasonable size of memory to work with?
If ($servermemory / $instances.count -lt ... ) { "too dangerous";...}